File: Matlab_code/calibrate_cupanemometer.m
%% Script used to calibrate individual cup anemometer measurements to wind speed
% Author: Jena Shields
% Last updated: 6/16/25
% Matlab version: MATLAB R2023b
% Done with a fan array and handheld anemometer 
% Each station is put in front of fan array at a particular wind speed,
% collect data for a few minutes. Upload the data into this script in
% batches by each wind speed. I did three wind speeds in total 

clear all
close all

%% wind speeds measured by handheld anemometer 
vs_fpm = [405,600,820]; %wind speed in feet per minute, change for your measurements
vs_mps = vs_fpm*0.00508; %wind speed in meters per second
%% Loop through txt files to get data 

%List of stations to be calibrated
ws_list = [3,4,5,6,7];
count2mps = zeros(length(vs_mps),length(ws_list)); % cup count to meters per second initialized variable
avg_ccps = zeros(length(vs_mps),length(ws_list)); %average cup count per second initialized variable


for k = 1:length(vs_mps)
    v_fpm = vs_fpm(k);
    v_mps = vs_mps(k);
    for i = 1:length(ws_list)
        num = num2str(ws_list(i)); %Select one weather station at a time, convert number to string
        %File name for station with this speed
        filename= strcat('sample_data\calibration_data\calibration_',num2str(v_fpm), 'fpm\weatherstation_', num, '_c.txt'); %File name for station with this speed
        
        %Open the .txt file, read it, and close it
        fid = fopen(filename, 'rt');  %Open the file to read it in text mode
        tline = 'temp_RTC,temp,humidity,wind_dir,cup_count'; %The order of measurements saved (this should match the arduino code order)
        headers = strsplit(tline, ',');     %a cell array of strings of the measurement names
        datacell = textscan(fid, '%s%s%f%f%f%f%f', 'Delimiter',',', 'EndOfLine',';', 'CollectOutput', 1 ); %Data is read from the text file, 2 strings, 5 floating point numbers, each value separated by a comma, each line separated by a ;
        fclose(fid); %Close the file
    
        % take text data and make table
        dates =  datacell{1}; %This is where the string values are stored (has date and time info)
        data = num2cell(datacell{2}); %This is where the numerical data are stored
    
        dates_temp = strcat(dates(:, 1), {' '}, dates(:, 2)); %Combine the two string arrays to make one date/time string array
        dates_in_dt = datetime(dates_temp, 'InputFormat', 'uu/MM/dd HH:mm:ss', 'Format','uu/MM/dd, HH:mm:ss'); %Turn into a datetime array in specified format
    
        data_table = cell2table(data, "VariableNames", headers); %Make a table with the numerical data with the specified variable names 
        final_table = table2timetable(data_table, "RowTimes", dates_in_dt);  %Form a time table with the numerical data and datetime array
    
        %Initialize new variables in the data table for:
        final_table.cupcountps = NaN(size(final_table,1), 1); %cup count per second
        final_table.cupcountp10s = NaN(size(final_table,1), 1); %cup count averaged over 10 seconds
        final_table.cupcountps(1) = final_table.cup_count(1); %The first value in cupcount per second the first measurement of cup_count from text file
    
        %Cup count per second calculation, takes into account the cup_count
        %variable resetting to zero regularly 
        for j = 2:length(final_table.cup_count)
            if final_table.cup_count(j) < final_table.cup_count(j-1) %If the next cup count value is less than the previous, cup count per second is undefined
                final_table.cupcountps(j) = NaN;
            else
                final_table.cupcountps(j) = final_table.cup_count(j)-final_table.cup_count(j-1); %If the next cup count value is greater than previous, cup count per second is the difference 
            end
        end
        
        %Cup count averaged over 10 seconds calculation, takes into account the
        %cup_count variable resetting to zero regularly 
        for j = 11:length(final_table.cup_count)
            if final_table.cup_count(j) < final_table.cup_count(j-10) %If the 10th next cup count value is less than the current, cup count per 10 seconds is undefined
                final_table.cupcountp10s(j) = NaN;
            else
                final_table.cupcountp10s(j) = (final_table.cup_count(j)-final_table.cup_count(j-10))/10; %If the 10th next cup count value is greater than current, cup count averaged over 10 seconds is the difference, divded by 10
            end
        end
    
        %Filter final_table raw values if outliers/anomalies 
        if v_fpm == 820
            if ws_list(i) == 5
                h = final_table.cupcountp10s(44:end); %first few data points abnormal
            else
                h = final_table.cupcountp10s;
            end
        elseif v_fpm == 600
            h = final_table.cupcountp10s;
        elseif v_fpm == 405
            if ws_list(i) == 5
                h = rmoutliers(final_table.cupcountp10s); %remove outliers 
            else
                h = final_table.cupcountp10s;
            end
        end

        avg_ccps(k,i) = mean(h,"omitmissing"); %Calculate average cup count per second for this station for this speed

        count2mps(k,i) = v_mps/avg_ccps(k,i); % mps/count/s, conversion from average cup count per second to counts per m/s

        %If these have outliers, filter h above and rerun
        plot(h)
        title(strcat('station ', num, 'ccps ', num2str(avg_ccps(i))))
        pause(1)
        
    end
end   

%% Make table 

 avg_ccps = transpose(avg_ccps); %transpose the array of average cup count per second values
 count2mps = transpose(count2mps); %transpose the array of count to m/s vlaues
 calibrations = [transpose(ws_list), avg_ccps, count2mps]; %Make an array with the first column being station number, then the cup count per second values, then the conversion values
 
 %Make a string array to name each variable in the calibrations array
 var_names = strings(1,width(calibrations));
 var_names(1) = 'station';

 %Loops through the different speeds to make labels for the calibrations
 %array
 for i = 1:length(vs_mps)
     var_names(1+i) = strcat('ccps_', num2str(vs_fpm(i)), 'fps');

     var_names(1+length(vs_mps)+i) = strcat('cc2mps_',num2str(vs_fpm(i)),'fps');
 end

 %Convert the calibrations array two a table with the variable names
 %compiled above
 calibrations = array2table(calibrations, "VariableNames",var_names);

%% Calculate the slope of the cup count per second/wind speed for calibration

figure()
slopes = zeros(height(calibrations),1); %Initialize slopes variable
for i = 1:height(calibrations)
    plot(vs_mps, [calibrations{i,2},calibrations{i,3},calibrations{i,4}]) %Plot the cup count per second versus wind speed 
    xlabel('Wind speed [mps]')
    ylabel('Cup counts per second')
    hold on

    slopes(i) = (820*0.00508-405*0.00508)/(calibrations{i,4}-calibrations{i,2}); %Calculate the slope for each station
end

calibrations.slopes = slopes; %Save the slope in the calibrations table(calibration between cup count to m/s)

%% Save the calibration data 

save("sample_data\calibrations.mat","calibrations","-mat")

File: Python_code/calibrate_cupanemometer.py
# -*- coding: utf-8 -*-
"""
 Script used to calibrate individual cup anemometer measurements to wind speed
 Author: Jena Shields
 Last updated: 7/22/26
 Python version: Python 3.14.2

 Translated from matlab version with help from claude ai
"""

"""
Python (NumPy/pandas) conversion of the MATLAB anemometer calibration script.

Conversion notes:
  - The .txt calibration files use ';' as the record/line separator and ','
    as the field separator within a record (MATLAB's textscan used
    'EndOfLine', ';'). This is NOT a normal CSV, so pandas.read_csv won't
    work directly -- the file is read as raw text and split manually.
  - table2timetable -> a pandas DataFrame with a DatetimeIndex.
  - MATLAB's datetime format 'uu/MM/dd HH:mm:ss' (2-digit year) ->
    Python's strptime format "%y/%m/%d %H:%M:%S".
  - rmoutliers (default "median" method: remove points more than 3 scaled
    MADs from the median, scale factor 1.4826) is reproduced by hand, since
    there's no direct NumPy/pandas equivalent.
  - mean(h, "omitmissing") -> np.nanmean(h)
  - array2table -> a pandas DataFrame with named columns.
  - Saved as .csv at the end (a calibration table is naturally tabular);
    swap for np.savez if you'd rather keep everything in .npz for
    consistency with the rest of your pipeline -- see the commented
    block at the bottom.
"""
#%% Import dependences
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from pathlib import Path

#%%
# ---------------------------------------------------------------------------
# Wind speeds measured by handheld anemometer
# ---------------------------------------------------------------------------
vs_fpm = np.array([405, 600, 820])       # wind speed, feet per minute -- change for your measurements
vs_mps = vs_fpm * 0.00508                # wind speed, meters per second


#%%
# ---------------------------------------------------------------------------
# Helper: read one calibration .txt file into a DataFrame
# ---------------------------------------------------------------------------
def read_calibration_file(filename):
    """
    Parses a calibration .txt file where records are separated by ';'
    and fields within a record are separated by ','. Each record is:
    date, time, temp_RTC, temp, humidity, wind_dir, cup_count
    """
    headers = ["temp_RTC", "temp", "humidity", "wind_dir", "cup_count"]

    with open(filename, "r") as f:
        content = f.read()

    records = [r.strip() for r in content.split(";") if r.strip()]

    dates = []
    rows = []
    for rec in records:
        fields = [x.strip() for x in rec.split(",")]
        date_str, time_str = fields[0], fields[1]
        nums = [float(x) for x in fields[2:7]]
        dt = pd.to_datetime(f"{date_str} {time_str}", format="%y/%m/%d %H:%M:%S")
        dates.append(dt)
        rows.append(nums)

    df = pd.DataFrame(rows, columns=headers, index=pd.DatetimeIndex(dates))
    return df


# ---------------------------------------------------------------------------
# Helper: MATLAB rmoutliers default ("median" method) -- remove points more
# than 3 scaled MADs (scale factor 1.4826) from the median
# ---------------------------------------------------------------------------
def rmoutliers_median(x, threshold=3.0):
    x = np.asarray(x, dtype=float)
    med = np.nanmedian(x)
    scaled_mad = 1.4826 * np.nanmedian(np.abs(x - med))
    keep = np.abs(x - med) <= threshold * scaled_mad
    return x[keep]


# ---------------------------------------------------------------------------
# Loop through txt files to get data
# ---------------------------------------------------------------------------
ws_list = [3, 4, 5, 6, 7]  # list of stations to be calibrated

count2mps = np.zeros((len(vs_mps), len(ws_list)))  # cup count -> m/s
avg_ccps = np.zeros((len(vs_mps), len(ws_list)))   # avg cup count per second

plt.ion()  # allow non-blocking plots so the loop can keep going

for k, (v_fpm, v_mps) in enumerate(zip(vs_fpm, vs_mps)):
    for i, station in enumerate(ws_list):
        num = str(station)  # station number as string

        filename = (
            Path("sample_data") / "calibration_data" / f"calibration_{v_fpm}fpm"
            / f"weatherstation_{num}_c.txt"
        )

        final_table = read_calibration_file(filename)
        cup_count = final_table["cup_count"].to_numpy()
        n = len(cup_count)

        # --- cup count per second (handles counter resets) ---
        cupcountps = np.full(n, np.nan)
        cupcountps[0] = cup_count[0]
        for j in range(1, n):
            if cup_count[j] < cup_count[j - 1]:
                cupcountps[j] = np.nan
            else:
                cupcountps[j] = cup_count[j] - cup_count[j - 1]
        final_table["cupcountps"] = cupcountps

        # --- cup count averaged over 10 seconds (handles counter resets) ---
        cupcountp10s = np.full(n, np.nan)
        for j in range(10, n):
            if cup_count[j] < cup_count[j - 10]:
                cupcountp10s[j] = np.nan
            else:
                cupcountp10s[j] = (cup_count[j] - cup_count[j - 10]) / 10
        final_table["cupcountp10s"] = cupcountp10s

        # --- filter raw values for known outliers/anomalies ---
        if v_fpm == 820:
            if station == 5:
                h = final_table["cupcountp10s"].to_numpy()[43:]  # first few points abnormal
            else:
                h = final_table["cupcountp10s"].to_numpy()
        elif v_fpm == 600:
            h = final_table["cupcountp10s"].to_numpy()
        elif v_fpm == 405:
            if station == 5:
                h = rmoutliers_median(final_table["cupcountp10s"].to_numpy())
            else:
                h = final_table["cupcountp10s"].to_numpy()

        avg_ccps[k, i] = np.nanmean(h)               # avg cup count/sec for this station/speed
        count2mps[k, i] = v_mps / avg_ccps[k, i]     # m/s per (count/s)

        # If these have outliers, filter h above and rerun
        plt.figure()
        plt.plot(h)
        plt.title(f"station {num} ccps {avg_ccps[k, i]}")
        plt.pause(1)

# %%
# ---------------------------------------------------------------------------
# Make table
# ---------------------------------------------------------------------------
avg_ccps_T = avg_ccps.T     # stations x speeds
count2mps_T = count2mps.T   # stations x speeds

calibrations_arr = np.column_stack([np.array(ws_list), avg_ccps_T, count2mps_T])

var_names = ["station"]
for v in vs_fpm:
    var_names.append(f"ccps_{v}fps")
for v in vs_fpm:
    var_names.append(f"cc2mps_{v}fps")

calibrations = pd.DataFrame(calibrations_arr, columns=var_names)

#%%
# ---------------------------------------------------------------------------
# Calculate the slope of cup count per second / wind speed for calibration
# ---------------------------------------------------------------------------
plt.figure()
slopes = np.zeros(len(calibrations))

for i in range(len(calibrations)):
    row = calibrations.iloc[i]
    plt.plot(vs_mps, [row["ccps_405fps"], row["ccps_600fps"], row["ccps_820fps"]])

    slopes[i] = (820 * 0.00508 - 405 * 0.00508) / (row["ccps_820fps"] - row["ccps_405fps"])

plt.xlabel("Wind speed [mps]")
plt.ylabel("Cup counts per second")

calibrations["slopes"] = slopes

#%%
# ---------------------------------------------------------------------------
# Save the calibration data
# ---------------------------------------------------------------------------
out_folder = Path("sample_data")
calibrations.to_csv(out_folder / "calibrations.csv", index=False)

# If you'd rather keep this in .npz format for consistency with the rest of
# the pipeline:
# np.savez(
#     out_folder / "calibrations.npz",
#     station=calibrations["station"].to_numpy(),
#     var_names=np.array(calibrations.columns),
#     data=calibrations.to_numpy(),
# )
# %%
